### PR TITLE
Feature: Max expressions

### PR DESCRIPTION
**Pull request by jlozano**
@breml I moved this to a separate PR since my git history on the old PR was corrupted.

## Summary
This PR adds an option for the user to specify the maximum number of "steps" the parser should take before failing with a user defined error.

## Notes
- A "step" is defined to be a call to `parser.parseExpr` since a single rule can contain an infinite loop, e.g `rule_infinite = rule_infinite`.
- Switch exprCnt to `uint64` since this makes more sense than an `int` and ensures that the `p.exprCnt > p.maxExprCnt` conditional always makes sense.
- The default value of `parser.maxExprCnt` is `math.MaxUint64` to avoid a conditional check.

### CODE REVIEW

File: builder/static_code.go [hunk 1]
```diff
@@ -91,9 +113,7 @@ func ParseFile(filename string, opts ...Option) (i interface{}, err error) {
 		return nil, err
 	}
 	defer func() {
-		if closeErr := f.Close(); closeErr != nil {
-			err = closeErr
-		}
```
Comment (breml) [inline builder/static_code.go]: This change would revert commit 7397997f31ddb24949618f11bab1588199703721, which is not the intention. Needs to be changed here (please regenerate all the generated files with `make`).

Comment (jlozano) [reply builder/static_code.go]: oops, will fix, not sure how that got in there

**Comment by jlozano** [comment]
@breml responded to comments. I removed the error message string. I know we discussed possibly exporting an interface to signal to the caller that the file had failed to parse in the specified number of steps, however I ultimately decided to use the error pattern that is employed for the other errors. Individual clients can determine how they want to export this error (if at all), the unit test gives an example of how to check for it.

**Comment by jlozano** [comment]
@breml will squash commits once this is ready to merge, seems to be causing some issues with my git history otherwise.

File: test/max_expr_cnt/max_expr_cnt_test.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package main
+
+import "testing"
+
+func TestMaxExprCnt(t *testing.T) {
+	_, err := Parse("", []byte("infinite parse"), MaxExpressions(5))
+	if err == nil {
+		t.Errorf("expected non nil error message for testing max expr cnt option.")
```
Comment (breml) [inline test/max_expr_cnt/max_expr_cnt_test.go]: `t.Errorf` already includes `t.Fail` according to [godoc](https://godoc.org/testing#T.Errorf). Please remove `t.Fail()` calls in this file.
Why do you use `t.Errorf` in the first place (why do you want to continue executing the test after an unexpected behavior was encountered)? I suggest to change the calls to `t.Fatalf`.

Comment (jlozano) [reply test/max_expr_cnt/max_expr_cnt_test.go]: Switched to `t.Fatalf`

**Pull request merged**